File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/K._Выведи_диапазон/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(Node* left, Node* right, int value) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>

using namespace std;


using namespace std;

void print_range(Node* root, int l, int r) {
    /*
    Time Complexity O(h + k)
    Memory Complexity O(h)
    */
    if (!root) {
        return;
    }
    if (root->value >= l) {
        print_range(root->left, l, r);
    }
    if (root->value >= l && root->value <= r) {
        std::cout << root->value << " ";
    }
    if (root->value <= r) {
        print_range(root->right, l, r);
    }
}

#ifndef REMOTE_JUDGE
void test() {
    Node node1({nullptr, nullptr, 2});
    Node node2({nullptr, &node1, 1});
    Node node3({nullptr, nullptr, 8});
    Node node4({nullptr, &node3, 8});
    Node node5({&node4, nullptr, 9});
    Node node6({&node5, nullptr, 10});
    Node node7({&node2, &node6, 5});
    print_range(&node7, 2, 8);
    // expected output: 2 5 8 8
}

int main() {
  test();
}
#endif

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/K._Выведи_диапазон/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, left=None, right=None, value=0):
            self.right = right
            self.left = left
            self.value = value


def print_range(node, l, r):
    """
    Time Complexity O(h + k)
    Memory Complexity O(h)
    """
    if not node:
        return
    if node.value >= l:
        print_range(node.left, l, r)
    if l <= node.value <= r:
        print(node.value, end=' ')
    if node.value <= r:
        print_range(node.right, l, r)


def test():
    node1 = Node(None, None, 2)
    node2 = Node(None, node1, 1)
    node3 = Node(None, None, 8)
    node4 = Node(None, node3, 8)
    node5 = Node(node4, None, 9)
    node6 = Node(node5, None, 10)
    node7 = Node(node2, node6, 5)
    print_range(node7, 2, 8)
    # expected output: 2 5 8 8


if __name__ == '__main__':
    test()